src/phoenix/experimental/evals/functions/classify.py

```diff
@@ -22,6 +23,119 @@
+
+        model (BaseEvalModel): An LLM model class.
+
+        rails (List[str]): A list of strings representing the possible output classes of the model's
+        predictions.
+
+        system_instruction (Optional[str], optional): An optional system message.
+
+        verbose (bool, optional): If True, prints detailed info to stdout such as model invocation
+        parameters and details about retries and snapping to rails. Default False.
+
+        use_function_calling_if_available (bool, default=True): If True, use function calling
+        (if available) as a means to constrain the LLM outputs.
+
+    Returns:
+        List[ClassificationResult]: A list of tuples representing the predicted class and the
+        explanation for the prediction for each record in the dataframe. The list should have
+        the same length as the input dataframe and its values should be the entries in the rails
+        argument or "NOT_PARSABLE" if the model's prediction could not be parsed.
+    """
+    use_openai_function_call = (
+        use_function_calling_if_available
+        and isinstance(model, OpenAIModel)
+        and model.supports_function_calling
+    )
+
+    # TODO: support explanation without function calling
+    if not use_openai_function_call:
+        raise ValueError(
+            "explanation is not currently available for models that don't "
+            "support OpenAI function calling"
+        )
+
+    model_kwargs = {}
+    if use_openai_function_call:
+        openai_function = _default_openai_function(rails, True)
+        model_kwargs.update(
+            {
+                "function_call": {"name": openai_function["name"]},
+                "functions": [openai_function],
+            }
+        )
+
+    eval_template = normalize_template(template)
+    prompts = map_template(dataframe, eval_template)
+    with set_verbosity(model, verbose) as verbose_model:
+        responses = verbose_model.generate(
+            prompts.to_list(), instruction=system_instruction, **model_kwargs
+        )
+    printif(verbose, f"Snapping {len(responses)} responses to rails: {rails}")
+    if not use_openai_function_call:
+        return [
+            ClassificationResult(
+                label=_snap_to_rail(response, rails, verbose=verbose),
+                explanation=None,  # TODO: support explanation without function calling
+            )
+            for response in responses
+        ]
+
+    results = []
+    for response in responses:
+        raw_string = response
+        explanation = None
+        try:
+            function_arguments = json.loads(response, strict=False)
+            raw_string = function_arguments.get(_RESPONSE) or ""
+            explanation = function_arguments.get(_EXPLANATION)
+        except json.JSONDecodeError:
+            pass
+        label = _snap_to_rail(raw_string, rails, verbose=verbose)
+        results.append(ClassificationResult(label=label, explanation=explanation))
```

Comment:
Not sure if this has impact in python runtime but this feels like wasteful effort to have a single code path to attempt to make the processing of non function calling responses be treated exactly the same way. It's also a bit confusing to a reader as you don't really understand why a json.loads is being attempted for non function calling.

suggestion: why not just have two response parsers - one for function calling and one for not. Seems cleaner and more future proof

Reply:
Oh I see sorry misread the code 👍